src/experimental/tools/download2go/apis/dash/dashTools.ts

```diff
@@ -0,0 +1,297 @@
+    (acc, buffer) => {
+      if (buffer.byteLength > 0) {
+        acc.newBuffer.set(buffer, acc.offset);
+        acc.offset += buffer.byteLength;
+        return acc;
+      }
+      return acc;
+    },
+    { newBuffer: new Uint8Array(totalLengthBuffers), offset: 0 }
+  );
+  return concatBuffer.newBuffer;
+}
+
+/**
+ * A tool function that is used to filter depending
+ * on a callback but also permit to emit immediately of the given subject is emiting.
+ *
+ * @param (value) => boolean - a callback that should return a boolean
+ * @param AsyncSubject - An AsyncSubject that wait a single emission
+ * @returns Observable source
+ *
+ */
+export function takeUntilFilter<V, T>(
+  shouldEmit: (value: [V, T]) => boolean,
+  pause$: AsyncSubject<void>
+) {
+  let i = 0;
+  return (source: Observable<[V, T]>): Observable<[V, T]> => {
+    return new Observable<[V, T]>(obs => {
+      const subscription = source.subscribe(
+        value => {
+          try {
+            pause$.subscribe(() => {
+              if (i === 0) {
+                obs.next(value);
+                i += 1;
+              }
+            });
+            if (shouldEmit(value)) {
+              obs.next(value);
+              return;
+            }
+          } catch (err) {
+            obs.error(err);
+          }
+        },
+        err => obs.error(err),
+        () => obs.complete()
+      );
+
+      return subscription;
+    });
+  };
+}
+
+/**
+ * An operator that take a Observable in parameter that
+ * is emitting a progress status. Then, decide when to emit next manifest.
+ *
+ * @param progress$ - An Observable under the form ProgressBarBuilder type
+ * @returns - An Observable
+ *
+ */
+export function emitEveryNth(
+  progress$: ReplaySubject<IProgressBarBuilderAbstract>,
+  utilsBuilder: IUtils,
+  contentID: string,
+  pause$: AsyncSubject<void>
+) {
+  return (
+    source$: Observable<ILocalManifestOnline>
+  ): Observable<IDownloadManagerOutput> => {
+    const { storeManifestEvery, emitter } = utilsBuilder;
+    return new Observable((obs: Observer<IDownloadManagerOutput>) => {
+      const progressFiltered$ = progress$.pipe(
+        mergeMap(({ progress, ...props }) =>
+          of({
+            progress:
+              progress > 98 && progress < 100 ? progress : Math.round(progress),
+            ...props,
+          })
+        ),
+        distinctUntilKeyChanged("progress")
+      );
+      return combineLatest([source$, progressFiltered$])
+        .pipe(
+          distinctUntilChanged(
+            (
+              [_, { progress: prevProgress }],
+              [__, { progress: currProgress }]
+            ) => prevProgress === currProgress
+          ),
+          tap(builder => {
+            const [, { status, progress, size }] = builder;
+            emitter.trigger("progress", {
+              contentID,
+              progress,
+              size,
+              status,
+            });
+          }),
+          takeUntilFilter(
```

Comment:
As discussed, it should not be needed

Reply:
We should use ```merge``` operator here, that merge the `pause$` and the the `builder$`.